### PR TITLE
Rework the BoFiN HEOM documentation

### PR DESCRIPTION
**Pull request by hodgestar**
**Description**
This add documentation of the new BoFiN-based HEOM solver added in #1601.

**Todo**
- [x] Rework the documentation structure.
- [x] Update the introduction
- [x] Update the bosonic example
- [x] Update the fermionic example
- [x] Add a short history of the HEOM implementation
- [x] Mention that the old HEOM implementation is available as `dl_heom`
- [x] Add a link to the HEOM notebooks.
- [x] Update `.run(..., ado_init=True)` to accept an instance of `HierarchyADOsState` since this is what `.run(..., ado_return=True)` returns. Add tests for `ado_init=True`.

**Related PRs**
- #1601
- #1726

**Changelog**
Add and rework the HEOM documentation.

### CODE REVIEW

File: doc/guide/heom/fermionic.rst [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+
+.. math::
+
+    f_F (x) = (\exp(x) + 1)^{-1}
+
+As in the bosonic case we can approximate this integral with a Matsubara or
+Padé expansion. For the Lorentzian bath the Padé expansion converges much
+more quickly, so we will calculate the Padé expansion coefficients here.
+
+The Padé decomposition approximates the Fermi distubition as:
```
Comment (nwlambert) [inline doc/guide/heom/fermionic.rst:352]: distribution

**Review by nwlambert — COMMENTED** [review]
just some comments and typo fixes!  looks really good!

File: doc/guide/heom/bosonic.rst [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+If ``ado_return=True`` is passed to ``.run(...)`` the full set of auxilliary
+density operators (ADOs) that make up the hierarchy at each time will be
+returned as ``.ado_states``. We will describe how to use these to determine
+other properties, such as system-bath currents, later in the fermionic guide
+(see :ref:`heom-determining-currents`).
+
+If one has a full set of ADOs from a previous call of ``.run(...)`` you may
+supply it as the initial state of the solver by calling
+``.run(result.ado_states[-1], tlist, ado_init=True)``.
+
```
[the remark this answers is not in the document]
Comment (hodgestar) [reply doc/guide/heom/bosonic.rst:149]: Fixing this is on my TODO list. I was going to fix it in a separate PR, but let me just fix it in this one.

File: doc/guide/heom/bosonic.rst [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+Multiple baths
+--------------
+
+The :class:`~qutip.nonmarkov.heom.HEOMSolver` supports having a system interact
+with multiple environments. All that is needed is to supply a list of baths
+instead of a singe bath.
+
+In the example below we calculate the evolution of a small system where
+each energy eigenstate of the system interacts with a separate bath. Such
+as arrangement can model, for example, Fenna–Matthews–Olson (FMO)
```
[the remark this answers is not in the document]
Comment (hodgestar) [reply doc/guide/heom/bosonic.rst]: Link added.

File: doc/guide/heom/bosonic.rst [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+
+Multiple baths
+--------------
+
+The :class:`~qutip.nonmarkov.heom.HEOMSolver` supports having a system interact
+with multiple environments. All that is needed is to supply a list of baths
+instead of a singe bath.
+
+In the example below we calculate the evolution of a small system where
+each energy eigenstate of the system interacts with a separate bath. Such
```
[the remark this answers is not in the document]
Comment (hodgestar) [reply doc/guide/heom/bosonic.rst]: I changed "each energy eigenstate" to "each basis state" -- does that seem more accurate?

File: doc/guide/heom/bosonic.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (hodgestar) [reply doc/guide/heom/bosonic.rst:149]: Test and support added.

**Comment by hodgestar** [comment]
@nwlambert Ready for a re-review!

**Review by nwlambert — APPROVED** [review]
looks good to me!

**Pull request merged**